Feature: Databricks compatibility

fivetran-catfritz opened the pull request:
Confirm the following files were correctly updated automatically:
- [x] `hooks/pre-command` line added
- [x] `pipeline.yml` section added
- [x] `sample.profiles.yml` catalog updated
- [x] `integration_tests/dbt_project.yml` section added (check for dupes though)
- [x] Version updates for `dbt_project.yml` and `integration_tests/dbt_project.yml` (automatically moved to next breaking change)
- [x] Changelog (automatically moved to next breaking change, but check the PR links were added correctly!)

Manual updates are:
- [x] README
- [x] `packages.yml` for transform packages
- [x] Incremental Models (see validation below)
- [x] Regen Docs
- [x] Local Databricks run and test passing
- [x] Buildkite passing

**Incremental Validation:**
- Two incrememental models:
  - iterable__events
  - int_iterable__list_user_unnest
  - (note int_iterable__custom_events is not in use)
- Used seed data to validate. 
  - Ran on new schema with some rows removed from the seeds. Result of the int models mentioned above. Note that I sorted in descending order for visibility, and note row counts:
![Screenshot 2023-06-23 at 4 27 16 PM](https://github.com/fivetran/dbt_iterable/assets/111930712/35eed128-9597-4b2e-8b77-4958e2e8cf5b)

  - Added the rows back in to the seeds and ran on the same schema. I made sure to do a normal run and not a full-refresh. Result:
![Screenshot 2023-06-23 at 4 29 46 PM](https://github.com/fivetran/dbt_iterable/assets/111930712/93f60d42-5862-4b1d-ac3a-9bd9e21705ba)

  - I feel confident that the incremental logic is working as it should.

**Model Updates Validation:**
- I had to add logic in `int_iterable__list_user_unnest` for Databricks/Spark. This model takes `email_list_ids`, which is a string list of ids, breaks out each id, and then creates a separate row for each id. The ids are caputured in new column `list_id`, and 
I wanted to validate the output was the same. This model produces the same values as the postgres version:
![Screenshot 2023-06-23 at 5 05 08 PM](https://github.com/fivetran/dbt_iterable/assets/111930712/0f3c362f-d2c5-467a-9d61-d02462a84ddd)

  - Note, there is a discrepancy with the other warehouse versions, but I opened [this issue](https://github.com/fivetran/dbt_iterable/issues/31) to address since it is not a Databricks issue.

fivetran-joemarkiewicz commented on models/intermediate/int_iterable__list_user_unnest.sql:
```diff
@@ -71,9 +71,13 @@ with user_history as (
+            {% if target.type == 'snowflake' %}
+            email_list_id.value
+            {% elif target.type in ('spark','databricks') %}
+            email_list_id.col
```
Can you explain why databricks and spark are using the `col` column? Where is this coming from?

fivetran-catfritz replied:
Yes. From [the docs](https://docs.databricks.com/sql/language-manual/functions/explode_outer.html#returns):
> The column produced by `explode_outer` of an array is named `col`. 

I was thinking that seemed reasonable based on how snowflake appears to do a similar thing on lines 74-75.

fivetran-joemarkiewicz replied:
Thanks for the explanation, that is what I assumed but just wanted to make sure. Looks good to me!

fivetran-joemarkiewicz reviewed (changes requested):
@fivetran-catfritz thanks for working through this! I have a few comments and requested changes in the code below and also in this response. Let me know if you have questions.

My additional are the following:
- Can we update the `iterable__events` model to leverage the same config and strategies as the `int_iterable__list_user_unnest` model. I feel it would be best to have a consistent strategy across the same package. 
  - Be sure to update the CHANGELOG with this specific update. 
- For the `insert_overwrite` strategy we will need to change the file_format to be `parquet` per the [docs](https://docs.getdbt.com/reference/resource-configs/spark-configs#the-insert_overwrite-strategy).
  - Before adding this I would want to confirm that this will have no effect on the other warehouse strategies (I don't believe it will, but would like to confirm).
  - Can we also confirm that this file format is available to all (or at least the majority) of Databricks users. If the file format is not available for all users, we should make this call out in the README. 
  - Right off the bat it looks like we will need to make a callout in the README that Databricks users connecting via ODBC driver will need to add a cluster Spark Config in order for this to work. (see the usage notes from the link above).

fivetran-catfritz commented:
- I have updated and tested `iterable__events`. 
- I updated to `file_format='parquet'` and updated the README. I didn't add language about ODBC because it looks like it's not an issue in the `dbt-databricks` adapter (the issue is for `dbt-spark`), but pointed to the docs if there were issues. Reference [dbt-databricks](https://docs.getdbt.com/reference/resource-configs/databricks-configs#the-insert_overwrite-strategy) vs [dbt-spark](https://docs.getdbt.com/reference/resource-configs/spark-configs#the-insert_overwrite-strategy)
- For parquet vs delta, there are pluses and minuses, but no hard issues I could find.

fivetran-joemarkiewicz commented:
> * I have updated and tested `iterable__events`.
> * I updated to `file_format='parquet'` and updated the README. I didn't add language about ODBC because it looks like it's not an issue in the `dbt-databricks` adapter (the issue is for `dbt-spark`), but pointed to the docs if there were issues. Reference [dbt-databricks](https://docs.getdbt.com/reference/resource-configs/databricks-configs#the-insert_overwrite-strategy) vs [dbt-spark](https://docs.getdbt.com/reference/resource-configs/spark-configs#the-insert_overwrite-strategy)
> * For parquet vs delta, there are pluses and minuses, but no hard issues I could find.

Thanks so much for making the updates and diving into the incremental nuances with the `insert_overwrite` strategy for databricks. I will review today and let you know if I have any other notes 😄

fivetran-joemarkiewicz reviewed (approved):
Thanks for making all the requested changes and diving into the nuances of incremental strategies for databricks! All looks good to me!
